networkx/algorithms/tests/test_summarization.py

```diff
@@ -213,10 +212,429 @@ def test_dedensify_edge_count(self):
         undirected graph
         """
         G = self.build_original_graph()
-        c_G, c_nodes = dedensify(G, threshold=2, copy=True)
+        c_G, c_nodes = nx.dedensify(G, threshold=2, copy=True)
         compressed_edge_count = len(c_G.edges())
         verified_original_edge_count = len(G.edges())
         assert compressed_edge_count <= verified_original_edge_count
         verified_compressed_G = self.build_compressed_graph()
         verified_compressed_edge_count = len(verified_compressed_G.edges())
         assert compressed_edge_count == verified_compressed_edge_count
+
+
+class TestSNAPEmpty:
+    graph_types = [nx.Graph, nx.DiGraph, nx.MultiGraph, nx.MultiDiGraph]
+
+    @pytest.mark.parametrize("graph_type", graph_types)
+    def test_summarization_empty(self, graph_type):
+        G = graph_type()
+        summary_graph = nx.snap_aggregation(G, node_attributes=("color",))
+        assert nx.is_isomorphic(summary_graph, G) is True
+
+
+class AbstractSNAP:
+    node_attributes = ("color",)
+
+    def test_summary_graph(self):
+        original_graph = self.build_original_graph()
+        summary_graph = self.build_summary_graph()
+
+        relationship_attributes = ("type",)
+        generated_summary_graph = nx.snap_aggregation(
+            original_graph,
+            self.node_attributes,
+            relationship_attributes,
+        )
+        relabeled_summary_graph = self.deterministic_labels(generated_summary_graph)
+        assert nx.is_isomorphic(summary_graph, relabeled_summary_graph)
+
+    def deterministic_labels(self, G):
+        node_labels = list(G.nodes)
+        node_labels = sorted(node_labels, key=lambda n: sorted(G.nodes[n]["group"])[0])
+        node_labels.sort()
+
+        label_mapping = dict()
+        for index, node in enumerate(node_labels):
+            label = "Supernode-%s" % index
+            label_mapping[node] = label
+
+        return nx.relabel_nodes(G, label_mapping)
+
+
+class TestSNAPNoEdgeTypes(AbstractSNAP):
+    relationship_attributes = ()
+
+    def test_summary_graph(self):
+        original_graph = self.build_original_graph()
+        summary_graph = self.build_summary_graph()
```

Comment:
This pattern is generally unsafe - calling methods that are only defined in child classes but not in the `AbstractSNAP`.